### PR TITLE
docs: add versioning reference

### PR DESCRIPTION
**Pull request by joshblack**
Add versioning reference for teams. This should explain what our policy is with respect to versioning and what types of changes in `carbon-components-react` correspond to which semver bumps

#### Changelog

**New**

- Add versioning.md guide 

**Changed**

**Removed**

### CODE REVIEW

File: docs/guides/versioning.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+> 2. **MINOR** version when you add functionality in a backwards compatible
+>    manner, and
+> 3. **PATCH** version when you make backwards compatible bug fixes.
+>
+> _Additional labels for pre-release and build metadata are available as
+> extensions to the MAJOR.MINOR.PATCH format._
+
+As a result, whenever you see a `minor` or `patch` update for a package from the
+Carbon Design System you should feel confident that you can update without
+anything breaking in project.
```
Comment (connor-leech) [inline docs/guides/versioning.md]:
```suggestion
anything breaking in your project.
```

File: docs/guides/versioning.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+**Note:** If you ever bring in an update and it does break something in your
+project, please create an
+[issue](https://github.com/carbon-design-system/carbon/issues/new?assignees=&labels=type%3A+bug+%F0%9F%90%9B&template=bug-report.md&title=Breaking%20change)
+and we will resolve the issue as quickly as possible.
+
+In the following sections, you'll find specific details for our packages and the
+types of changes you can expect to occur as a consumer of that package. We'll
+try and highlight common changes in code and the corresponding `semver` bump
```
Comment (connor-leech) [inline docs/guides/versioning.md]:
```suggestion
try to highlight common changes in code and the corresponding `semver` bump
```

File: docs/guides/versioning.md [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+| An existing prop is removed                                         | `major`     |
+| An existing prop type is changed to be more specific                | `major`     |
+| An existing prop type is changed to be more generic                 | `minor`     |
+| A `PropTypes.func` prop type is changed to have different arguments | `major`     |
+| A `PropTypes.func` prop type is changed to have additional argument | `minor`     |
+| A `PropTypes.func` prop type is changed to have fewer arguments     | `major`     |
+
+### Examples
+
+#### A new prop is added to a component
```
Comment (connor-leech) [inline docs/guides/versioning.md:71]:
```suggestion
#### A new prop is added to a component:
```

We've finally decided that we *are* doing colons after H4s

Comment (joshblack) [reply docs/guides/versioning.md:71]: As this isn't for the website, do we need to follow this convention?

File: docs/guides/versioning.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  );
+}
+
+ExampleComponent.propTypes = {
+  propA: PropTypes.string,
++  propB: PropTypes.string,
+};
+```
+
+#### An existing prop is deprecated
```
Comment (connor-leech) [inline docs/guides/versioning.md:94]:
```suggestion
#### An existing prop is deprecated:
```

File: docs/guides/versioning.md [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+ExampleComponent.propTypes = {
+  propA: PropTypes.string,
+-  propB: PropTypes.string,
++  propB: deprecate(PropTypes.string),
+};
+```
+
+#### An existing prop is removed
```
Comment (connor-leech) [inline docs/guides/versioning.md:115]:
```suggestion
#### An existing prop is removed:
```

File: docs/guides/versioning.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+  );
+}
+
+ExampleComponent.propTypes = {
+  propA: PropTypes.string,
+-  propB: deprecate(PropTypes.string),
+};
+```
+
+#### An existing prop type is changed to be more specific
```
Comment (connor-leech) [inline docs/guides/versioning.md:138]:
```suggestion
#### An existing prop type is changed to be more specific:
```

File: docs/guides/versioning.md [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+ExampleComponent.propTypes = {
+  propA: PropTypes.string,
+-  propB: PropTypes.node,
++  propB: PropTypes.string,
+};
+```
+
+#### An existing prop type is changed to be more generic
```
Comment (connor-leech) [inline docs/guides/versioning.md:159]:
```suggestion
#### An existing prop type is changed to be more generic:
```

**Review by abbeyhrt — APPROVED** [review]
Looks good! Definitely made minor/major and patch release changes more clear to me 🎉

File: docs/guides/versioning.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+#### A `PropTypes.func` prop type is changed to have different arguments
+
+semver bump: **major**
+
+```diff
+import { SomeComponent } from 'carbon-components-react';
+
+function ExampleComponent() {
+  function onChange(arg) {
+    // If a or b change types (from number to string) or if they are removed
```
Comment (emyarod) [inline docs/guides/versioning.md:189]: should there be a diff in this example? even just for the comments

**Review by emyarod — APPROVED** [review]

File: docs/guides/versioning.md (hunk 8 above)
Comment (joshblack) [reply docs/guides/versioning.md:189]: I'd be down! Have any suggestions?

**Review by vpicone — COMMENTED** [review]
Should we add a note irt properties being added/taken away from arguments? Should adding a new property to an object be considered a breaking change?

**Comment by joshblack** [comment]
@vpicone I'd be down! Feel free to commit any/all examples, would love more details

**Comment by vpicone** [comment]
@joshblack Not sure if it's worth adding, what do you think? Something like 'we reserve the right to add new properties to callbacks' so don't do anything fishy like map over properties or something.

```js
const onClick = () => {
  onClickProp({a: 'foo', b:'foo'})

  // property added ✅ minor release?
  onClickProp({a: 'foo', b:'foo', c: 'new property'}) 

  // property removed ❌ major release
  onClickProp({a: 'foo' }) 
}
```

Even though it's not a change in primitive type, it could be kind of a gray area. Not sure this matters when we normalize these event handlers across components.

**Comment by joshblack** [comment]
@vpicone I say go for it, better to have more examples! Feel free to add onto this one, or I can merge and you can do another PR.

Only nit would be for styles that we use:

```js
function onClick() {
  // ...
}

// Or even
function handleOnClick() {
  onClick({ a: 'string', b: 0 });
  // ...
}
```

Just for consistency in the file 👍

**Pull request merged**